libraries/chain/include/eosio/chain/combined_database.hpp

```diff
@@ -14,77 +36,88 @@
-         kv_undo_stack     = src.kv_undo_stack;
-         src.kv_undo_stack = nullptr;
-         return *this;
+      combined_session make_session() {
+         return backing_store == backing_store_type::ROCKSDB ? combined_session(db, kv_undo_stack)
+                                                             : combined_session(db);
       }
 
-      void push() {
-         try {
-            try {
-               if (cb_session)
-                  cb_session->push();
-               cb_session    = nullptr;
-               kv_undo_stack = nullptr;
-            }
-            FC_LOG_AND_RETHROW()
-         }
-         CATCH_AND_EXIT_DB_FAILURE()
-      }
+      void set_revision(uint64_t revision);
 
-      void squash() {
-         try {
-            try {
-               if (cb_session)
-                  cb_session->squash();
-               if (kv_undo_stack)
-                  kv_undo_stack->squash(false);
-               cb_session    = nullptr;
-               kv_undo_stack = nullptr;
-            }
-            FC_LOG_AND_RETHROW()
-         }
-         CATCH_AND_EXIT_DB_FAILURE()
-      }
+      void undo();
 
-      void undo() {
-         try {
-            try {
-               if (cb_session)
-                  cb_session->undo();
-               if (kv_undo_stack)
-                  kv_undo_stack->undo(false);
-               cb_session    = nullptr;
-               kv_undo_stack = nullptr;
-            }
-            FC_LOG_AND_RETHROW()
-         }
-         CATCH_AND_EXIT_DB_FAILURE()
-      }
+      void commit(int64_t revision);
+
+      void flush();
+
+      std::unique_ptr<kv_context> create_kv_context(name receiver, kv_resource_manager resource_manager,
+                                                    const kv_database_config& limits);
+
+      void add_to_snapshot(const eosio::chain::snapshot_writer_ptr& snapshot, const eosio::chain::block_state& head,
+                           const eosio::chain::authorization_manager&                    authorization,
+                           const eosio::chain::resource_limits::resource_limits_manager& resource_limits) const;
+
+      void read_from_snapshot(const snapshot_reader_ptr& snapshot, uint32_t blog_start, uint32_t blog_end,
+                              backing_store_type backing_store, eosio::chain::authorization_manager& authorization,
+                              eosio::chain::resource_limits::resource_limits_manager& resource_limits,
+                              eosio::chain::fork_database& fork_db, eosio::chain::block_state_ptr& head,
+                              uint32_t& snapshot_head_block, const eosio::chain::chain_id_type& chain_id);
+
+    private:
+      void add_contract_tables_to_snapshot(const snapshot_writer_ptr& snapshot) const;
+      void read_contract_tables_from_snapshot(const snapshot_reader_ptr& snapshot);
+
+      backing_store_type       backing_store;
+      chainbase::database&     db;
+      b1::chain_kv::database   kv_database;
```

Comment:
If you utilized an `std::variant` for the db and kv_db then you could do away with the explicit backing_store_type and variable.

Reply:
As Brian said above, chainbase is used even when KV entries are stored in chain_kv.